coccinelle/heap_freetuple.cocci

```diff
@@ -0,0 +1,29 @@
+// find heap_form_tuple with missing heap_freetuple calls
+@ heap_form_tuple @
+expression tuple;
+position p;
+@@
+
+tuple@p = heap_form_tuple(...);
+
+@safelist@
+expression tuple;
+position heap_form_tuple.p;
+@@
+
+tuple@p = heap_form_tuple(...);
+...
+(
+return tuple;
+|
+heap_freetuple(tuple)
+|
+HeapTupleGetDatum(tuple)
+)
+@depends on !safelist@
+expression tuple;
+position heap_form_tuple.p;
```

Comment:
OK, I think I kinda understand what it does after looking at it for a couple of minutes. Maybe some high-level comment about this language would help, since it's not the main language we use.

By the way, it looks like we first match `@heap_form_tuple@`, then `@safelist@`, then `@depends on !safelist@`? Does the first match add anything, e.g. can we instead match `@safelist@` and then `@depends on !safelist@`?

Also, position in `@depends on !safelist@` is `heap_form_tuple.p`, but then we use just `p` in the diff, is it intentional?

Reply:
I'll add a readme with some short intro. The first block is the anchor which gets referenced in the safelist and the output block.  inside the @@ is the block header which has variable declarations which can reference other blocks. Inside the diff part of the block you can't namespace variables as that is set up in the block header.